core/src/environment.rs

```diff
@@ -101,35 +103,18 @@ impl<K: Hash + Eq, V: PartialEq> Environment<K, V> {
     /// Creates an iterator that visits all layers from the most recent one to the oldest.
     /// The element iterator type is `Rc<HashMap<K, V>>`.
     pub fn iter_layers(&self) -> EnvLayerIter<'_, K, V> {
-        EnvLayerIter {
-            env: if !self.was_cloned() {
-                Some(NonNull::from(self))
-            } else {
-                // if was cloned, current is the same as first of previous (that cannot be empty
-                // then)
-                self.previous
-                    .borrow()
-                    .as_ref()
-                    // SAFETY: created from Rc, so cannot be null
-                    .map(|prev| unsafe { NonNull::new_unchecked(Rc::as_ptr(prev) as *mut _) })
-            },
-            _marker: PhantomData,
-        }
+        EnvLayerIter { env: Some(self) }
     }
 
     /// Creates an iterator that visits all elements from the Environment, from the oldest layer to
     /// the most recent one. It uses this order, so calling `collect` on this iterator to create a
     /// hashmap would have the same values as the Environment. The element iterator type is `(&'env
     /// K, &'env V)`, with `'env` being the lifetime of the Environment.
-    pub fn iter_elems(&self) -> EnvElemIter<'_, K, V> {
-        let mut env: Vec<NonNull<HashMap<K, V>>> = self
-            .iter_layers()
-            // SAFETY: Rc::as_ptr never returnes null
-            .map(|hmap| unsafe { NonNull::new_unchecked(Rc::as_ptr(hmap) as *mut _) })
-            .collect();
-        // SAFETY: by design, env cannot be empty, and coming from an Rc, it is well aligned and initialized
-        let current_map = unsafe { env.pop().unwrap().as_ref() }.iter();
-        EnvElemIter { env, current_map }
+    pub fn iter_elems(
+        &self,
+    ) -> std::iter::Flatten<std::iter::Rev<std::vec::IntoIter<&HashMap<K, V>>>> {
+        let env: Vec<&HashMap<K, V>> = self.iter_layers().map(|hmap| &**hmap).collect();
```

Comment:
Do we need to `collect` and then call `into_iter` right away for borrowing reasons (need to consume the first iterator)?

Reply:
I think it's not so much for borrowing reasons as because we want to visit in reverse order. But the collecting was in the previous version also.

I wonder if it would be faster to implement `iter_elems` by walking the environment from newest to oldest, and keeping a `HashSet` to avoid repeating any keys. That seems maybe faster than the current approach that requires two collects. I'll try it.

Reply:
Ok, I tried optimizing `iter` but it turns out that `iter` isn't actually used anywhere so I deleted it instead.

Before deleting the new `iter`, I tried using it to replace the existing uses of `iter_elems`, but it didn't help. Either way there's an allocation, and I guess the `Vec` for going backwards over the hashmaps beats the `HashSet` for tracking repeated elements.